<!DOCTYPE html>
<html>
<head>
<style>
* {
  box-sizing: border-box;
}

.row {
  margin-left:-5px;
  margin-right:-5px;
}
  
.column {
  float: left;
  width: 50%;
  padding: 5px;
}

/* Clearfix (clear floats) */
.row::after {
  content: "";
  clear: both;
  display: table;
}

table {
  border-collapse: collapse;
  border-spacing: 0;
  width: 100%;
  border: 1px solid #ddd;
}

th, td {
  text-align: left;
  padding: 16px;
}

tr:nth-child(even) {
  background-color: #f2f2f2;
}
</style>
</head>
<body>

<h2 style="text-align: center" >BBM409 : Introduction to Machine Learning Lab.</h2>
<h4 style="text-align: center">Assignment 1 :  K-Nearest Neighbor Algorithm.</h4>
<br>
<div class="row">
  <div class="column">
    <table>
      <tr>
        <th>First Name</th>
        <th>Last Name</th>
        <th>No</th>
      </tr>
      <tr>
        <td>Esad </td>
        <td>Boran</td>
        <td>21827206</td>
      </tr>
      <tr>
        <td>Evren</td>
        <td>Çağılcı</td>
        <td>21945977</td>
      </tr>
    </table>
  </div>
</div>

</body>
</html>

# 1. Introduction

K-Nearest Neighbor algorithm was utilized in this experiment to solve the classification and regression problems. The K-Nearest Neighbors algorithm calculates the likelihood that a data point will join one group or another based on which group the data points closest to it are members of. The input for classification and regression consists of a data set's k closest training instances.  This assignment is divided into two sections. A KNN-classification experiment is included in the first section, and a KNN-regression experiment is included in the second.

# 2. Background

## 2.1  Classification Performance Metric

"Accuracy", "Precision", and "Recall" are calculated to test the model's success.

Accuracy rate indicates how accurate the model finds the classes.
<img src="Accuracy ss.png" style = "width:500px;height:150px"/>

Precision is the fraction of relevant instances among the retrieved instances.
<img src="Precision ss.png" style = "width:300px;height:120px"/>

Recall is the fraction of relevant instances that were retrieved.
<img src="Recall ss.png" style = "width:300px;height:100px"/>

## 2.2 Regression Performance Metric

A measure of mistakes between paired observations describing the same phenomenon is called mean absolute error (MAE).
<img src="MAE ss.png" style = "width:700px;height:60px"/>

## 2.3 Min-Max Scaler

MinMax Scaling is a case where data takes values between 0 and 1. Here the distribution is similar to the distribution of the data.
<img src="Feature Normalization ss.png" style = "width:300px;height:125px"/>

Where, ni represents the ith normalized element of your specific feature column (f) and fi represents the ith original element of your specific feature column (f).

## 2.4 results_cv() method

This method generates y_predict values according to K values of K-NN and k values of k-fold Cross-Validation. After this part, this method calculates the model's accuracy, precision and recall ratios with using the y_predict values which was generated before.

# 3 Required Libraries

In [1]:

import pandas as pd 
import numpy as np
from KNN_model import *
from requiredFunc import *
from Calculations import *

# 4 Assignment Solution

#    4.1 Part I: Personality Classification

In [2]:
# We got the data with the pandas library
data = pd.read_csv("subset_16P.csv")
data.head()

Response Id  You regularly make new friends.  \
0        35874                               -1   
1        42624                                0   
2        55199                                0   
3        52983                                0   
4        22864                                0   

   You spend a lot of your free time exploring various random topics that pique your interest  \
0                                                  0                                            
1                                                  0                                            
2                                                  0                                            
3                                                  0                                            
4                                                  0                                            

   Seeing other people cry can easily make you feel like you want to cry too  \
0                                                 -1                           
1                                                  1                           
2                                                 -2                           
3                                                  0                           
4                                                  2                           

   You often make a backup plan for a backup plan.  \
0                                                1   
1                                                0   
2                                               -1   
3                                                1   
4                                                1   

   You usually stay calm, even under a lot of pressure  \
0                                                 -1     
1                                                  0     
2                                                  2     
3                                                 -2     
4                                                  0     

   At social events, you rarely try to introduce yourself to new people and mostly talk to the ones you already know  \
0                                                 -2                                                                   
1                                                  0                                                                   
2                                                 -2                                                                   
3                                                 -1                                                                   
4                                                 -2                                                                   

   You prefer to completely finish one project before starting another.  \
0                                                 -2                      
1                                                 -1                      
2                                                  0                      
3                                                  0                      
4                                                 -1                      

   You are very sentimental.  \
0                          0   
1                          0   
2                          0   
3                          0   
4                          0   

   You like to use organizing tools like schedules and lists.  ...  \
0                                                 -1           ...   
1                                                  0           ...   
2                                                 -1           ...   
3                                                  1           ...   
4                                                  1           ...   

   You believe that pondering abstract philosophical questions is a waste of time.  \
0                                                  0                                 


In [3]:
# We converted the dataframe to np.array to process easier on the data using the numpy library.
arr_data = np.array(data)

In [4]:
# We have generated the shuff() method to prevent data in the dataset from being collected in certain places.
# This method randomly replaces the elements in the dataset.
shuff_data = shuff(arr_data)

In [5]:
# There are 16 different Personality attribute in type_labels. This will be used to calculate recall and precise.
type_labels = pd.unique(data.Personality)
type_labels

array(['ENTP', 'INTP', 'ESTP', 'ENFJ', 'ISTJ', 'ENFP', 'INFP', 'INTJ',
       'ESTJ', 'ESFJ', 'ENTJ', 'ISFJ', 'INFJ', 'ISFP', 'ESFP', 'ISTP'],
      dtype=object)

## 4.1.1 Data preprocessing and separation

In [6]:
# 60 questions and their answers.
X = np.array(shuff_data[:,1:-1])

In [7]:
# The labels of data.
y = np.array(shuff_data[:,-1])

In [8]:
# We used normalize() function to apply min-max scaling to the dataset.
X_normalized = normalize(X)

## 4.1.2 Unweighted KNN Classification

### 4.1.2.1 Data Without Normalization

In [9]:
# The explanation of results_cv() function is at the background section. 
# Accuracy, precision, and recall ratios of unweighted and non-normalized K-NN model.
unweighted_without_norm = results_cv(X,y,model_type="Classification",weights=False,normalized = False,calculations="metric")
unweighted_without_norm # unweighted_without_norm = [accuracy table , precision table , recall table]

[             1       3       5       7       9
 Fold-1  0.9765  0.9845  0.9860  0.9860  0.9860
 Fold-2  0.9785  0.9870  0.9885  0.9885  0.9890
 Fold-3  0.9755  0.9850  0.9865  0.9875  0.9875
 Fold-4  0.9740  0.9875  0.9880  0.9895  0.9905
 Fold-5  0.9750  0.9875  0.9885  0.9885  0.9895,
                1         3         5         7         9
 Fold-1  0.976526  0.984415  0.986069  0.986046  0.986117
 Fold-2  0.978417  0.986916  0.988276  0.988350  0.988793
 Fold-3  0.975349  0.984908  0.986390  0.987394  0.987434
 Fold-4  0.974033  0.987528  0.987902  0.989445  0.990367
 Fold-5  0.975222  0.987628  0.988737  0.988737  0.989684,
                1         3         5         7         9
 Fold-1  0.975735  0.984062  0.985683  0.985586  0.985675
 Fold-2  0.978272  0.987198  0.988687  0.988675  0.989183
 Fold-3  0.975414  0.984952  0.986567  0.987607  0.987359
 Fold-4  0.974549  0.987898  0.988300  0.989669  0.990605
 Fold-5  0.975373  0.987675  0.988648  0.988648  0.989681]

### 4.1.2.2 Data With Normalization

In [10]:
# Accuracy, precision, and recall ratios of unweighted and normalized K-NN model.
unweighted_with_norm = results_cv(X,y,model_type="Classification",weights=False,normalized = True,calculations="metric")
unweighted_with_norm # unweighted_with_norm = [accuracy table , precision table , recall table]

[             1       3       5       7       9
 Fold-1  0.9670  0.9775  0.9825  0.9845  0.9840
 Fold-2  0.9645  0.9835  0.9865  0.9890  0.9890
 Fold-3  0.9600  0.9745  0.9825  0.9855  0.9825
 Fold-4  0.9615  0.9790  0.9845  0.9845  0.9865
 Fold-5  0.9645  0.9820  0.9860  0.9855  0.9860,
                1         3         5         7         9
 Fold-1  0.967030  0.977774  0.982603  0.984572  0.984088
 Fold-2  0.964094  0.983469  0.986276  0.988793  0.988741
 Fold-3  0.959620  0.974231  0.982601  0.985371  0.982419
 Fold-4  0.961550  0.978924  0.984528  0.984410  0.986466
 Fold-5  0.964579  0.982493  0.986281  0.985855  0.986363,
                1         3         5         7         9
 Fold-1  0.966729  0.977569  0.982070  0.984023  0.983757
 Fold-2  0.964531  0.983653  0.986668  0.989150  0.989107
 Fold-3  0.959516  0.973882  0.982362  0.985427  0.982284
 Fold-4  0.962535  0.979237  0.984504  0.984684  0.986471
 Fold-5  0.964686  0.981903  0.986126  0.985721  0.986146]

## 4.1.3 Weighted KNN Classification

### 4.1.3.1 Data Without Normalization

In [11]:
# Accuracy, precision, and recall ratios of weighted and non-normalized K-NN model.
weighted_without_norm = results_cv(X,y,model_type="Classification",weights=True,normalized = False,calculations="metric")
weighted_without_norm # weighted_without_norm = [accuracy table , precision table , recall table]

[             1       3       5       7       9
 Fold-1  0.9765  0.9850  0.9855  0.9860  0.9865
 Fold-2  0.9785  0.9880  0.9885  0.9885  0.9885
 Fold-3  0.9755  0.9855  0.9885  0.9880  0.9875
 Fold-4  0.9740  0.9885  0.9885  0.9905  0.9905
 Fold-5  0.9750  0.9865  0.9885  0.9885  0.9895,
                1         3         5         7         9
 Fold-1  0.976526  0.984919  0.985526  0.986046  0.986597
 Fold-2  0.978417  0.987833  0.988276  0.988276  0.988276
 Fold-3  0.975349  0.985414  0.988402  0.987894  0.987434
 Fold-4  0.974033  0.988447  0.988406  0.990367  0.990367
 Fold-5  0.975222  0.986637  0.988737  0.988737  0.989684,
                1         3         5         7         9
 Fold-1  0.975735  0.984406  0.985125  0.985586  0.986135
 Fold-2  0.978272  0.988081  0.988687  0.988687  0.988687
 Fold-3  0.975414  0.985323  0.988516  0.988066  0.987341
 Fold-4  0.974549  0.988730  0.988714  0.990605  0.990605
 Fold-5  0.975373  0.986668  0.988648  0.988648  0.989681]

### 4.1.3.2 Data With Normalization

In [12]:
# Accuracy, precision, and recall ratios of weighted and normalized K-NN model.
weighted_with_norm = results_cv(X,y,model_type="Classification",weights=True,normalized = True,calculations="metric")
weighted_with_norm # weighted_with_norm = [accuracy table , precision table , recall table]

[             1       3       5       7       9
 Fold-1  0.9670  0.9795  0.9830  0.9845  0.9845
 Fold-2  0.9645  0.9860  0.9880  0.9885  0.9885
 Fold-3  0.9600  0.9780  0.9825  0.9845  0.9825
 Fold-4  0.9615  0.9805  0.9850  0.9850  0.9865
 Fold-5  0.9645  0.9805  0.9850  0.9850  0.9850,
                1         3         5         7         9
 Fold-1  0.967030  0.979692  0.983169  0.984595  0.984592
 Fold-2  0.964094  0.985931  0.987848  0.988365  0.988225
 Fold-3  0.959620  0.977603  0.982465  0.984413  0.982419
 Fold-4  0.961550  0.980404  0.984991  0.984860  0.986466
 Fold-5  0.964579  0.980908  0.985305  0.985352  0.985409,
                1         3         5         7         9
 Fold-1  0.966729  0.979057  0.982489  0.984050  0.984025
 Fold-2  0.964531  0.986072  0.988126  0.988614  0.988637
 Fold-3  0.959516  0.977396  0.982279  0.984318  0.982266
 Fold-4  0.962535  0.980290  0.984981  0.985134  0.986471
 Fold-5  0.964686  0.980618  0.985098  0.985214  0.985192]

## 4.1.4 Performance Metric

### 4.1.4.1 Accuracy Analysis

In [13]:
data_accuracy = [unweighted_with_norm[0], unweighted_without_norm[0], weighted_with_norm[0],weighted_without_norm[0]]
data_accuracy = pd.concat(data_accuracy,axis=1)
data_accuracy.columns = [x for x in range(1,21)]
data_accuracy.to_csv('data_accuracy.csv', index=False) 
data_accuracy # In the data_accuracy table,#1-5 ----> Unweighted without normalization for k = 1,3,5,7,9
                                           #6-10----> Unweighted with normalization    for k = 1,3,5,7,9
                                           #11-15---> Weighted with normalization      for k = 1,3,5,7,9
                                           #16-20---> Weighted without normalization   for k = 1,3,5,7,9

1       2       3       4       5       6       7       8   \
Fold-1  0.9670  0.9775  0.9825  0.9845  0.9840  0.9765  0.9845  0.9860   
Fold-2  0.9645  0.9835  0.9865  0.9890  0.9890  0.9785  0.9870  0.9885   
Fold-3  0.9600  0.9745  0.9825  0.9855  0.9825  0.9755  0.9850  0.9865   
Fold-4  0.9615  0.9790  0.9845  0.9845  0.9865  0.9740  0.9875  0.9880   
Fold-5  0.9645  0.9820  0.9860  0.9855  0.9860  0.9750  0.9875  0.9885   

            9       10      11      12      13      14      15      16  \
Fold-1  0.9860  0.9860  0.9670  0.9795  0.9830  0.9845  0.9845  0.9765   
Fold-2  0.9885  0.9890  0.9645  0.9860  0.9880  0.9885  0.9885  0.9785   
Fold-3  0.9875  0.9875  0.9600  0.9780  0.9825  0.9845  0.9825  0.9755   
Fold-4  0.9895  0.9905  0.9615  0.9805  0.9850  0.9850  0.9865  0.9740   
Fold-5  0.9885  0.9895  0.9645  0.9805  0.9850  0.9850  0.9850  0.9750   

            17      18      19      20  
Fold-1  0.9850  0.9855  0.9860  0.9865  
Fold-2  0.9880  0.9885  0.9885  0.9885  
Fold-3  0.9855  0.9885  0.9880  0.9875  
Fold-4  0.9885  0.9885  0.9905  0.9905  
Fold-5  0.9865  0.9885  0.9885  0.9895

In [14]:
# accuracy_avarage is a pandas dataFrame which includes mean values of accuracy ratios.
accuracy_avarage = data_accuracy.mean()
accuracy_avarage = pd.DataFrame(data_accuracy.mean()).T
accuracy_avarage.index = ["Average of Folds"]
accuracy_avarage.to_csv('accuracy_avarage.csv', index=False)
accuracy_avarage

1       2       3       4       5       6       7   \
Average of Folds  0.9635  0.9793  0.9844  0.9858  0.9856  0.9759  0.9863   

                      8      9       10      11      12      13      14  \
Average of Folds  0.9875  0.988  0.9885  0.9635  0.9809  0.9847  0.9855   

                      15      16      17      18      19      20  
Average of Folds  0.9854  0.9759  0.9867  0.9879  0.9883  0.9885

### 4.1.4.2 Precision Analysis

In [15]:
data_precision = [unweighted_with_norm[1], unweighted_without_norm[1], weighted_with_norm[1],weighted_without_norm[1]]
data_precision = pd.concat(data_precision,axis=1)
data_precision.columns = [x for x in range(1,21)]
data_precision.to_csv('data_precision.csv', index=False)
data_precision # In the data_precision table,#1-5 ----> Unweighted without normalization for k = 1,3,5,7,9
                                             #6-10----> Unweighted with normalization    for k = 1,3,5,7,9
                                             #11-15---> Weighted with normalization      for k = 1,3,5,7,9
                                             #16-20---> Weighted without normalization   for k = 1,3,5,7,9

1         2         3         4         5         6         7   \
Fold-1  0.967030  0.977774  0.982603  0.984572  0.984088  0.976526  0.984415   
Fold-2  0.964094  0.983469  0.986276  0.988793  0.988741  0.978417  0.986916   
Fold-3  0.959620  0.974231  0.982601  0.985371  0.982419  0.975349  0.984908   
Fold-4  0.961550  0.978924  0.984528  0.984410  0.986466  0.974033  0.987528   
Fold-5  0.964579  0.982493  0.986281  0.985855  0.986363  0.975222  0.987628   

              8         9         10        11        12        13        14  \
Fold-1  0.986069  0.986046  0.986117  0.967030  0.979692  0.983169  0.984595   
Fold-2  0.988276  0.988350  0.988793  0.964094  0.985931  0.987848  0.988365   
Fold-3  0.986390  0.987394  0.987434  0.959620  0.977603  0.982465  0.984413   
Fold-4  0.987902  0.989445  0.990367  0.961550  0.980404  0.984991  0.984860   
Fold-5  0.988737  0.988737  0.989684  0.964579  0.980908  0.985305  0.985352   

              15        16        17        18        19        20  
Fold-1  0.984592  0.976526  0.984919  0.985526  0.986046  0.986597  
Fold-2  0.988225  0.978417  0.987833  0.988276  0.988276  0.988276  
Fold-3  0.982419  0.975349  0.985414  0.988402  0.987894  0.987434  
Fold-4  0.986466  0.974033  0.988447  0.988406  0.990367  0.990367  
Fold-5  0.985409  0.975222  0.986637  0.988737  0.988737  0.989684

In [16]:
# precision_avarage is a pandas dataFrame which includes mean values of precision ratios.
precision_average = data_precision.mean()
precision_average = pd.DataFrame(data_precision.mean()).T
precision_average.index = ["Average of Folds"]
precision_average.to_csv('precision_average.csv', index=False)
precision_average

1         2         3       4         5        6   \
Average of Folds  0.963375  0.979378  0.984458  0.9858  0.985615  0.97591   

                        7         8         9         10        11        12  \
Average of Folds  0.986279  0.987475  0.987994  0.988479  0.963375  0.980908   

                        13        14        15       16       17        18  \
Average of Folds  0.984756  0.985517  0.985422  0.97591  0.98665  0.987869   

                        19        20  
Average of Folds  0.988264  0.988472

### 4.1.4.3 Recall Analysis

In [17]:
data_recall = [unweighted_with_norm[2], unweighted_without_norm[2], weighted_with_norm[2],weighted_without_norm[2]]
data_recall = pd.concat(data_recall,axis=1)
data_recall.columns = [x for x in range(1,21)]
data_recall.to_csv('data_recall.csv', index=False)
data_recall # In the data_recall table,#1-5 ----> Unweighted without normalization for k = 1,3,5,7,9
                                       #6-10----> Unweighted with normalization    for k = 1,3,5,7,9
                                       #11-15---> Weighted with normalization      for k = 1,3,5,7,9
                                       #16-20---> Weighted without normalization   for k = 1,3,5,7,9

1         2         3         4         5         6         7   \
Fold-1  0.966729  0.977569  0.982070  0.984023  0.983757  0.975735  0.984062   
Fold-2  0.964531  0.983653  0.986668  0.989150  0.989107  0.978272  0.987198   
Fold-3  0.959516  0.973882  0.982362  0.985427  0.982284  0.975414  0.984952   
Fold-4  0.962535  0.979237  0.984504  0.984684  0.986471  0.974549  0.987898   
Fold-5  0.964686  0.981903  0.986126  0.985721  0.986146  0.975373  0.987675   

              8         9         10        11        12        13        14  \
Fold-1  0.985683  0.985586  0.985675  0.966729  0.979057  0.982489  0.984050   
Fold-2  0.988687  0.988675  0.989183  0.964531  0.986072  0.988126  0.988614   
Fold-3  0.986567  0.987607  0.987359  0.959516  0.977396  0.982279  0.984318   
Fold-4  0.988300  0.989669  0.990605  0.962535  0.980290  0.984981  0.985134   
Fold-5  0.988648  0.988648  0.989681  0.964686  0.980618  0.985098  0.985214   

              15        16        17        18        19        20  
Fold-1  0.984025  0.975735  0.984406  0.985125  0.985586  0.986135  
Fold-2  0.988637  0.978272  0.988081  0.988687  0.988687  0.988687  
Fold-3  0.982266  0.975414  0.985323  0.988516  0.988066  0.987341  
Fold-4  0.986471  0.974549  0.988730  0.988714  0.990605  0.990605  
Fold-5  0.985192  0.975373  0.986668  0.988648  0.988648  0.989681

In [18]:
# recall_avarage is a pandas dataFrame which includes mean values of recall ratios.
recall_average = data_recall.mean()
recall_average = pd.DataFrame(data_recall.mean()).T
recall_average.index = ["Average of Folds"]
recall_average.to_csv('recall_average.csv', index=False)
recall_average

1         2         3         4         5         6   \
Average of Folds  0.963599  0.979249  0.984346  0.985801  0.985553  0.975869   

                        7         8         9         10        11        12  \
Average of Folds  0.986357  0.987577  0.988037  0.988501  0.963599  0.980687   

                        13        14        15        16        17        18  \
Average of Folds  0.984594  0.985466  0.985318  0.975869  0.986642  0.987938   

                        19       20  
Average of Folds  0.988319  0.98849

# 4.2 Part 2: Energy Efficiency Estimation 

In [19]:
# We got the data with the pandas library
data = pd.read_csv("energy_efficiency_data.csv")
data

Relative_Compactness  Surface_Area  Wall_Area  Roof_Area  Overall_Height  \
0                    0.98         514.5      294.0     110.25             7.0   
1                    0.98         514.5      294.0     110.25             7.0   
2                    0.98         514.5      294.0     110.25             7.0   
3                    0.98         514.5      294.0     110.25             7.0   
4                    0.90         563.5      318.5     122.50             7.0   
..                    ...           ...        ...        ...             ...   
763                  0.64         784.0      343.0     220.50             3.5   
764                  0.62         808.5      367.5     220.50             3.5   
765                  0.62         808.5      367.5     220.50             3.5   
766                  0.62         808.5      367.5     220.50             3.5   
767                  0.62         808.5      367.5     220.50             3.5   

     Orientation  Glazing_Area  Glazing_Area_Distribution  Heating_Load  \
0              2           0.0                          0         15.55   
1              3           0.0                          0         15.55   
2              4           0.0                          0         15.55   
3              5           0.0                          0         15.55   
4              2           0.0                          0         20.84   
..           ...           ...                        ...           ...   
763            5           0.4                          5         17.88   
764            2           0.4                          5         16.54   
765            3           0.4                          5         16.44   
766            4           0.4                          5         16.48   
767            5           0.4                          5         16.64   

     Cooling_Load  
0           21.33  
1           21.33  
2           21.33  
3           21.33  
4           28.28  
..            ...  
763         21.40  
764         16.88  
765         17.11  
766         16.61  
767         16.03  

[768 rows x 10 columns]

In [20]:
# We converted the dataframe to np.array to process easier on the data using the numpy library.
arr_data = np.array(data)

In [21]:
# We have generated the shuff() method to prevent data in the dataset from being collected in certain places.
# This method randomly replaces the elements in the dataset.
arr_data = shuff(arr_data)

# 4.2.1  Data preprocessing

In [22]:
# 8 attributes and their values.
X = np.array(arr_data[:,0:-2])

In [23]:
# Heating Load values.
y_1= np.array(arr_data[:,-2])

In [24]:
# Cooling Load values.
y_2= np.array(arr_data[:,-1])

In [25]:
# We used normalize() function to apply min-max scaling to the dataset.
X_normalized = normalize(X)

# 4.2.2 For Heating Load

## 4.2.2.1 Unweighted KNN Regression

### 4.2.2.1.1  Data Without Normalization

In [26]:
# The explanation of results_cv() function is at the background section. 
# Mean absolute error values of unweighted and non-normalized K-NN model.
unweighted_without_norm = results_cv(X,y_1,model_type="Regression",weights=False,normalized = False,calculations="mae")
unweighted_without_norm

mae_1     mae_3     mae_5     mae_7     mae_9
Fold-1  2.783137  1.762745  1.575542  1.760971  2.003181
Fold-2  2.539150  1.861547  1.581778  1.894099  2.127110
Fold-3  2.720392  1.996645  1.797137  2.059991  2.300160
Fold-4  2.517190  1.742789  1.576105  1.812110  2.009971
Fold-5  2.575817  1.682789  1.346928  1.615425  1.830167

### 4.2.2.1.2 Data With Normalization

In [27]:
# Mean absolute error values of unweighted and normalized K-NN model.
unweighted_with_norm = results_cv(X,y_1,model_type="Regression",weights=False,normalized = True,calculations="mae")
unweighted_with_norm

mae_1     mae_3     mae_5     mae_7     mae_9
Fold-1  3.251699  2.060806  2.102627  2.071373  1.989172
Fold-2  2.761503  1.694684  1.775647  1.755201  1.785679
Fold-3  3.050392  1.930022  2.132549  1.923380  1.896333
Fold-4  3.239739  1.992048  1.941399  1.800299  1.718054
Fold-5  2.902288  1.880414  1.761190  1.623483  1.588751

## 4.2.2.2 Weighted KNN Regression 

### 4.2.2.2.1 Data Without Normalization

In [28]:
# Mean absolute error values of Weighted and non-normalized K-NN model.
weighted_without_norm = results_cv(X,y_1,model_type="Regression",weights=True,normalized = False,calculations="mae")
weighted_without_norm 

mae_1     mae_3     mae_5     mae_7     mae_9
Fold-1  2.783137  2.308170  2.151581  2.165530  2.264518
Fold-2  2.539150  2.370014  2.239597  2.298630  2.397051
Fold-3  2.720392  2.571465  2.401256  2.482147  2.579980
Fold-4  2.517190  2.333381  2.170519  2.210798  2.275698
Fold-5  2.575817  2.219771  2.010954  2.043859  2.126408

### 4.2.2.2.2 Data With Normalization

In [29]:
# Mean absolute error values of Weighted and normalized K-NN model.
weighted_with_norm = results_cv(X,y_1,model_type="Regression",weights=True,normalized = True,calculations="mae")
weighted_with_norm

mae_1     mae_3     mae_5     mae_7     mae_9
Fold-1  3.251699  2.018520  2.057860  2.045901  1.977570
Fold-2  2.761503  1.648122  1.694012  1.693220  1.724396
Fold-3  3.050392  1.875185  2.039672  1.897488  1.861147
Fold-4  3.239739  1.986022  1.944673  1.827054  1.738369
Fold-5  2.902288  1.823006  1.747813  1.632449  1.585135

## 4.2.2.3 Performance Analysis of  Mean Absolute Error for Heating Load 

In [30]:
data_mae_heating_load = [unweighted_with_norm, unweighted_without_norm, weighted_with_norm,weighted_without_norm]
data_mae_heating_load = pd.concat(data_mae_heating_load,axis=1)
data_mae_heating_load.columns = [x for x in range(1,21)]
data_mae_heating_load.to_csv('data_mae_heating_load.csv', index=False) 
data_mae_heating_load # In the data_maheating load table,
                                           #1-5 ----> Unweighted without normalization for k = 1,3,5,7,9
                                           #6-10----> Unweighted with normalization    for k = 1,3,5,7,9
                                           #11-15---> Weighted with normalization      for k = 1,3,5,7,9
                                           #16-20---> Weighted without normalization   for k = 1,3,5,7,9

1         2         3         4         5         6         7   \
Fold-1  3.251699  2.060806  2.102627  2.071373  1.989172  2.783137  1.762745   
Fold-2  2.761503  1.694684  1.775647  1.755201  1.785679  2.539150  1.861547   
Fold-3  3.050392  1.930022  2.132549  1.923380  1.896333  2.720392  1.996645   
Fold-4  3.239739  1.992048  1.941399  1.800299  1.718054  2.517190  1.742789   
Fold-5  2.902288  1.880414  1.761190  1.623483  1.588751  2.575817  1.682789   

              8         9         10        11        12        13        14  \
Fold-1  1.575542  1.760971  2.003181  3.251699  2.018520  2.057860  2.045901   
Fold-2  1.581778  1.894099  2.127110  2.761503  1.648122  1.694012  1.693220   
Fold-3  1.797137  2.059991  2.300160  3.050392  1.875185  2.039672  1.897488   
Fold-4  1.576105  1.812110  2.009971  3.239739  1.986022  1.944673  1.827054   
Fold-5  1.346928  1.615425  1.830167  2.902288  1.823006  1.747813  1.632449   

              15        16        17        18        19        20  
Fold-1  1.977570  2.783137  2.308170  2.151581  2.165530  2.264518  
Fold-2  1.724396  2.539150  2.370014  2.239597  2.298630  2.397051  
Fold-3  1.861147  2.720392  2.571465  2.401256  2.482147  2.579980  
Fold-4  1.738369  2.517190  2.333381  2.170519  2.210798  2.275698  
Fold-5  1.585135  2.575817  2.219771  2.010954  2.043859  2.126408

In [31]:
# data_mae_heating_load_average is a pandas dataFrame which includes mean values of mean absolute error values.
data_mae_heating_load_average = data_mae_heating_load.mean()
data_mae_heating_load_average = pd.DataFrame(data_mae_heating_load.mean()).T
data_mae_heating_load_average.index = ["Average of Folds"]
data_mae_heating_load_average.to_csv('data_mae_heating_load_average.csv', index=False)
data_mae_heating_load_average

1         2         3         4         5         6   \
Average of Folds  3.041124  1.911595  1.942682  1.834747  1.795598  2.627137   

                        7         8         9         10        11        12  \
Average of Folds  1.809303  1.575498  1.828519  2.054118  3.041124  1.870171   

                        13        14        15        16       17        18  \
Average of Folds  1.896806  1.819222  1.777323  2.627137  2.36056  2.194781   

                        19        20  
Average of Folds  2.240193  2.328731

# 4.2.3 For Cooling Load

## 4.2.3.1 Unweighted KNN Regression

### 4.2.3.1.1 Data Without Normalization

In [32]:
# Mean absolute error values of unweighted and non-normalized K-NN model.
unweighted_without_norm = results_cv(X,y_2,model_type="Regression",weights=False,normalized = False,calculations="mae")
unweighted_without_norm

mae_1     mae_3     mae_5     mae_7     mae_9
Fold-1  2.242484  1.484031  1.405294  1.562568  1.719419
Fold-2  2.139542  1.387538  1.307817  1.589673  1.791910
Fold-3  2.206797  1.634662  1.585516  1.798627  2.000356
Fold-4  2.013987  1.556885  1.588719  1.797852  1.951264
Fold-5  2.129150  1.386688  1.235268  1.438095  1.643471

### 4.2.3.1.2 Data With Normalization

In [33]:
# Mean absolute error values of unweighted and normalized K-NN model.
unweighted_with_norm = results_cv(X,y_2,model_type="Regression",weights=False,normalized = True,calculations="mae")
unweighted_with_norm

mae_1     mae_3     mae_5     mae_7     mae_9
Fold-1  3.533007  2.370087  2.355869  2.306321  2.238860
Fold-2  3.052418  1.927211  2.029686  1.990504  1.965541
Fold-3  3.268758  2.103377  2.253935  2.068992  2.047589
Fold-4  3.580523  2.248475  2.214013  1.991531  1.876093
Fold-5  3.221895  1.969150  1.961111  1.890700  1.726224

## 4.2.3.2 Weighted KNN Regression 

### 4.2.3.2.1 Data Without Normalization

In [34]:
# Mean absolute error values of Weighted and non-normalized K-NN model.
weighted_without_norm = results_cv(X,y_2,model_type="Regression",weights=True,normalized = False,calculations="mae")
weighted_without_norm 

mae_1     mae_3     mae_5     mae_7     mae_9
Fold-1  2.242484  1.773669  1.633130  1.630768  1.722173
Fold-2  2.139542  1.763704  1.679503  1.736659  1.831516
Fold-3  2.206797  2.016653  1.882798  1.946534  2.027887
Fold-4  2.013987  1.849166  1.792846  1.837078  1.903729
Fold-5  2.129150  1.777224  1.610122  1.639911  1.710596

### 4.2.3.2.2 Data With Normalization 

In [35]:
# Mean absolute error values of Weighted and normalized K-NN model.
weighted_with_norm = results_cv(X,y_2,model_type="Regression",weights=True,normalized = True,calculations="mae")
weighted_with_norm

mae_1     mae_3     mae_5     mae_7     mae_9
Fold-1  3.533007  2.335698  2.325811  2.302533  2.243091
Fold-2  3.052418  1.918139  1.981594  1.954060  1.941657
Fold-3  3.268758  2.079761  2.191652  2.062891  2.042345
Fold-4  3.580523  2.245959  2.207516  2.050980  1.945411
Fold-5  3.221895  1.948704  1.953266  1.904497  1.775247

## 4.2.3.3 Performance Analysis of Mean Absolute Error for Cooling Load

In [36]:
data_mae_cooling_load = [unweighted_with_norm, unweighted_without_norm, weighted_with_norm,weighted_without_norm]
data_mae_cooling_load = pd.concat(data_mae_cooling_load,axis=1)
data_mae_cooling_load.columns = [x for x in range(1,21)]
data_mae_cooling_load.to_csv('data_mae_cooling_load.csv', index=False) 
data_mae_cooling_load # In the data_maheating load table,
                                           #1-5 ----> Unweighted without normalization for k = 1,3,5,7,9
                                           #6-10----> Unweighted with normalization    for k = 1,3,5,7,9
                                           #11-15---> Weighted with normalization      for k = 1,3,5,7,9
                                           #16-20---> Weighted without normalization   for k = 1,3,5,7,9

1         2         3         4         5         6         7   \
Fold-1  3.533007  2.370087  2.355869  2.306321  2.238860  2.242484  1.484031   
Fold-2  3.052418  1.927211  2.029686  1.990504  1.965541  2.139542  1.387538   
Fold-3  3.268758  2.103377  2.253935  2.068992  2.047589  2.206797  1.634662   
Fold-4  3.580523  2.248475  2.214013  1.991531  1.876093  2.013987  1.556885   
Fold-5  3.221895  1.969150  1.961111  1.890700  1.726224  2.129150  1.386688   

              8         9         10        11        12        13        14  \
Fold-1  1.405294  1.562568  1.719419  3.533007  2.335698  2.325811  2.302533   
Fold-2  1.307817  1.589673  1.791910  3.052418  1.918139  1.981594  1.954060   
Fold-3  1.585516  1.798627  2.000356  3.268758  2.079761  2.191652  2.062891   
Fold-4  1.588719  1.797852  1.951264  3.580523  2.245959  2.207516  2.050980   
Fold-5  1.235268  1.438095  1.643471  3.221895  1.948704  1.953266  1.904497   

              15        16        17        18        19        20  
Fold-1  2.243091  2.242484  1.773669  1.633130  1.630768  1.722173  
Fold-2  1.941657  2.139542  1.763704  1.679503  1.736659  1.831516  
Fold-3  2.042345  2.206797  2.016653  1.882798  1.946534  2.027887  
Fold-4  1.945411  2.013987  1.849166  1.792846  1.837078  1.903729  
Fold-5  1.775247  2.129150  1.777224  1.610122  1.639911  1.710596

In [37]:
# data_mae_cooling_load_average is a pandas dataFrame which includes mean values of mean absolute error values.
data_mae_cooling_load_average = data_mae_cooling_load.mean()
data_mae_cooling_load_average = pd.DataFrame(data_mae_cooling_load.mean()).T
data_mae_cooling_load_average.index = ["Average of Folds"]
data_mae_cooling_load_average.to_csv('data_mae_cooling_load_average.csv', index=False)
data_mae_cooling_load_average

1        2         3        4         5         6   \
Average of Folds  3.33132  2.12366  2.162923  2.04961  1.970861  2.146392   

                        7         8         9         10       11        12  \
Average of Folds  1.489961  1.424523  1.637363  1.821284  3.33132  2.105652   

                        13        14       15        16        17       18  \
Average of Folds  2.131968  2.054992  1.98955  2.146392  1.836083  1.71968   

                       19       20  
Average of Folds  1.75819  1.83918

# 5 Error Analysis

## 5.1 Error Analysis for Part 1

 In Part 1, all for Accuracy, Precision and Recall ratios, 5-Fold Cross Validation gives the best ratio.
 For K values, when K = 9 , our model gives the best ratio. (9 Nearest Neighbor)
 Weighted KNN models gives better ratios then Unweighted KNN models.
 Non-normalized KNN models gives better ratios then Normalized KNN models.
 
 Weighted KNN models gives better ratios because in weighted models, we consider not only nearest neighbors. We also consider how close they are. Weighted KNN models work better when the dataset is large.
 
 Non-normalized KNN models gives better ratios because our data's ranges are not large.(between -3 and +3) Normalized KNN models work better when data's ranges are large.
 
 Time analysis:
 
 In KNN.fit_transform(), we calculated the distances just one time to decrease the execution time. Thanks to this, we can say that our predict code works in just a few seconds. Total execution time for 10.000 sized dataset is approximately 35 minutes for 4 models.(Weighted/Unweighted , Normalized/Non-Normalized)
 The total execution time decreases to approximately 1 minute if we shrink the data size to 1000. The reason of this is calculating the distances are working in O(n^2).

## 5.2 Error Analysis for Part 2

In Part 2 for both Heating Load and Cooling Load, 1-Fold Cross Validation gives the minimum mean absolute error(mea). For K values, when K = 5, our model gives the estimation rate.(5 Nearest Neighbor) The reason that K = 5, is because our data is small.(768 sample) If our dataset was larger, K value might be higher for best estimation rate.

Unweighted KNN Model gives better estimation ratio then Weighted KNN Model. The reason for this is our dataset is small.
Normalized KNN Model gives better estimatiın ratio then Non-normalized KNN Model. The reason for this is our data's ranges are large.

Time analysis:

Our code works less than 1 minute because dataset is too small.

# 6 Conclusion

The fundamentals of the KNN algorithm were covered. 
We noted that the KNN algorithm uses weighted and weightles, two distinct estimate methods. 
We looked into the significance of the K value. 
We investigated the differences and solutions to the issue by putting the KNN method to the test on both regression and classification situations. 
On our data, we used min-max scaling algorithm, then we looked at the outcomes. 
We have chosen a course of action for creating several machine learning models.

# 7.References

<ul>
  <li>https://numpy.org/doc/stable/index.html</li>
  <li>https://www.datacamp.com/tutorial/k-nearest-neighbor-classification-scikit-learn</li>
  <li>https://stackoverflow.com</li>
  <li>BBM406 Lecture Notes</li>
</ul>